### ISSUE TITLE
Offline Client session not found  (Session doesn't have required client invalid_grant)

### ISSUE THREAD

**Issue by imashwin-v** [keycloak/keycloak#21402]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

oidc

### Describe the bug

We are facing error when trying to get an access token via offline token using refresh_token grant in the token end point.

`Session doesn't have required client` is error thrown.

Deep diving into this issue we can see there are two sessions involved
1) OfflineUser session
2) OfflineClient session

Both these sessions have their own caches and entities.

**Entities:** offline_user_session, offline_client_session
**Cache:** offlineSessions, offlineClientSessions

When we use the offline token to get the access token, there is a method `findOfflineUserSession` in `TokenManager` which finds the matching session either from DB or cache. 

If the session is not found in the cache, a method to fetch both user session and client session from the DB is called and the session is returned for further validation and gives us an access token if session is valid.

If the session is found in the cache, then it is returned for further validation. Here only one cache(offlineSession) is queried to check the session, where as client session (offlineClientSession) cache is not checked. 

Hence if the clientSession cache is empty, the session validation fails at `userSession.getAuthenticatedClientSessionByClient(client.getId())` in the TokenManager class and fails with ,

`{
    "error": "invalid_grant",
    "error_description": "Session doesn't have required client"
}`

### Version

19.0.3

### Expected behavior

One should be able to get access token using refresh token with scope offline access if the token is present in the cache or in the DB (when enabled lazy loading)

### Actual behavior

`Session doesn't have required client` is error thrown.

`{
    "error": "invalid_grant",
    "error_description": "Session doesn't have required client"
}`

### How to Reproduce?

We are using Keycloak 19.0.3 with Wildfly.

Below is the configuration to set to distributed cache in-order to test this in a clustered environment where we face this error often due to cache rebalancing issues found in Infinispan during restart of the pods in the cluster.

**OfflineSession Cache**
`/subsystem=infinispan/cache-container=keycloak/distributed-cache=offlineSessions/memory=heap:write-attribute(name=max-entries, value=100)`

**OfflineClientSessions Cache**
`/subsystem=infinispan/cache-container=keycloak/distributed-cache=offlineClientSessions/memory=heap:write-attribute(name=max-entries, value=1)`

**Note:** It is important to set the value for offlineClient Session to be 1 to repro this issue easily. By default in Keycloak there are no max entries set for any caches, but in our environments we introduced these. 
In our environments all these entries are configurable according to the memory usage and our needs. 

Steps, 

1) Get an access and refresh token with scope offline access for any user via the token endpoint using password grant
2) Now exchange that offline token to get another access token for the same user using refresh_token grant, you might see no error and should receive the token successfully.
3) Now Get another access and refresh refresh token with scope offline access for any user via the token endpoint using password grant
4) Repeat the step 2 and try to get the access token again for the first tried user using refresh_token grant with first obtained offline_token. Now you will see the error 
`{
    "error": "invalid_grant",
    "error_description": "Session doesn't have required client"
}`

Here this error occurs since we don't find the required client session from the offline client session cache.
Since the offlineClientSession cache value is set to 1, it will replace old session entries with new entries in the cache for every offline token request and when a replaced old entry's offline token is tried to be exchanged  for access token, the above error will be thrown. 
The same happens in step 4, since step 3 will replace the step 1's entry.

### Anything else?

I have checked the Keycloak's latest code [21.1.2] and still see there isn't any change in fetching the offline tokens.
So even though i have tested this in a slightly older Keycloak version (19.0.3), the issue should still be relevant to the latest versions as well.

Similar issue,
Session doesn't have required client invalid_grant - https://github.com/keycloak/keycloak/issues/20392

**Comment by sschu** [comment]
I think this has been fixed with this PR: https://github.com/keycloak/keycloak/pull/17490
So can you please check with the current version?

**Comment by imashwin-v** [comment]
@sschu  Hi, we back-ported this change to our Keycloak (19.0.3) and it works, we were unable to test in the latest keycloak version but anyway it should work. Thanks.

**Comment by sschu** [comment]
Alright, closing this.

**Comment by ditogam** [comment]
Hello team,
I'm facing problem with key refresh token.

We deployed docker image *keycloak:22.0.5* , *SSO Session Idle* & *SSO Session* Max are configured 30 days, *Access Token Lifespan* for 10 minutes.

We do not use Infinitspan, docker command is :

> start-dev --http-enabled=true --hostname-strict=false --proxy reencrypt --http-relative-path=/auth

when Spring Gateway invokes grant_type "refresh_token" with refresh token, keycloak returns http 500 with message [invalid_grant] Session doesn't have required client

in docker we see only events like this

```
> 
> 2024-02-28 17:12:55,773 WARN  [org.keycloak.events] (executor-thread-192) type=REFRESH_TOKEN_ERROR, realmId=transguard, clientId=tg-bls, userId=ea131e0f-2570-4e92-a52f-a13427d629f7, ipAddress=10.116.225.8, error=invalid_token, grant_type=refresh_token, refresh_token_type=Offline, refresh_token_id=5e68c56e-0b1d-44e8-846e-bf24c2584eed, client_auth_method=client-secret
```


This happenes nearly every 30 min, if we refresh browser, everything works normally

As I know this docker was updated from keycloak version 13, but postgresql DB stayes the same

I have exactly the same image installed on other server and with the same config (also image was upgraded from v13), but there is no such issue.

Thanks for advices in advance.
